### PR TITLE
State cache behind feature flag

### PR DESCRIPTION
**Pull request by tudor-da**
Enable the mutable contract state cache behind a feature flag.
* Add cache size CLI configuration

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala [hunk 1]
```diff
@@ -280,12 +281,12 @@ class MutableCacheBackedContractStore(
...
   type EventSequentialId = Long
 
-  def apply(
+  private[cache] def apply(
       contractsReader: LedgerDaoContractsReader,
       signalNewLedgerHead: Offset => Unit,
       metrics: Metrics,
-      maxContractsCacheSize: Long = 100000L,
-      maxKeyCacheSize: Long = 100000L,
+      maxContractsCacheSize: Long = 100000,
```
Comment (mziolekda) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala:288]: I have come to appreciate the way that the defaults are handled in the `ParticipantIndexerConfig` case with the source of truth close to the code in question

File: ledger/participant-integration-api/src/main/scala/platform/apiserver/ApiServerConfig.scala [hunk 2]
```diff
@@ -28,4 +28,7 @@ case class ApiServerConfig(
     managementServiceTimeout: Duration,
     // TODO append-only: remove after removing support for the current (mutating) schema
     enableAppendOnlySchema: Boolean,
+    maxContractStateCacheSize: Long,
+    maxContractKeyStateCacheSize: Long,
+    enableMutableContractStateCacheConfig: Boolean,
```
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/platform/apiserver/ApiServerConfig.scala:33]: nit: Maybe just `enableMutableContractStateCache`?

File: ledger/participant-integration-api/src/main/scala/platform/index/ReadOnlySqlLedger.scala [hunk 3]
```diff
@@ -135,9 +141,18 @@ private[platform] object ReadOnlySqlLedger {
...
         lfValueTranslationCache: LfValueTranslationCache.Cache,
         contractsReader: LedgerDaoContractsReader,
-    ): Resource[ContractStore] =
-      TranslationCacheBackedContractStore
-        .owner(lfValueTranslationCache, contractsReader)
+    )(implicit executionContext: ExecutionContext): Resource[ContractStore] =
+      if (enableMutableContractStateCache)
+        MutableCacheBackedContractStore.owner(
+          contractsReader = contractsReader,
+          signalNewLedgerHead = _ => (), // TODO mutable contract state cache: wire up dispatcher
```
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/platform/index/ReadOnlySqlLedger.scala:148]: Are you planning to do this in this PR or a follow-up?

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala [hunk 4]
```diff
@@ -296,6 +297,25 @@ object MutableCacheBackedContractStore {
       ContractsStateCache(maxContractsCacheSize, metrics),
     )
 
+  def owner(
+      contractsReader: LedgerDaoContractsReader,
+      signalNewLedgerHead: Offset => Unit,
+      metrics: Metrics,
+      maxContractsCacheSize: Long = 100000L,
```
Comment (gerolf-da) [inline ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala:304]: Is it necessary to define the defaults here, considering they are defined elsewhere as well?

File: ledger/participant-integration-api/src/main/scala/platform/index/ReadOnlySqlLedger.scala (hunk 3 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/index/ReadOnlySqlLedger.scala:148]: We've decided to implement this in a follow-up PR in order to allow E2E integration now.

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala (hunk 4 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala:304]: You're right, they're redundant. Bubbled them up in the CLI config

File: ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/store/cache/MutableCacheBackedContractStore.scala]: Removed the defaults from here and bubbled them up in the CLI config.

File: ledger/sandbox/src/main/scala/platform/sandboxnext/Runner.scala [hunk 5]
```diff
@@ -236,6 +236,9 @@ class Runner(config: SandboxConfig) extends ResourceOwner[Port] {
                   seeding = config.seeding.get,
                   managementServiceTimeout = config.managementServiceTimeout,
                   enableAppendOnlySchema = false,
+                  maxContractStateCacheSize = 0L,
```
Comment (mziolekda) [inline ledger/sandbox/src/main/scala/platform/sandboxnext/Runner.scala:239]: Can we leave a TODO comment here to augment the settings for both the append-only and mutable-cache at the same time?

**Review by mziolekda — APPROVED** [review]
I have left one minor comment. I would appreciate if it were resolved before merging. Otherwise it looks good.

File: ledger/sandbox/src/main/scala/platform/sandboxnext/Runner.scala (hunk 5 above)
Comment (tudor-da) [reply ledger/sandbox/src/main/scala/platform/sandboxnext/Runner.scala:239]: Done

Comment (tudor-da) [reply ledger/sandbox/src/main/scala/platform/sandboxnext/Runner.scala:240]: Done

File: ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/ParticipantConfig.scala [hunk 6]
```diff
@@ -33,4 +35,7 @@ object ParticipantConfig {
 
   // this pool is used for all data access for the ledger api (command submission, transaction service, ...)
   val DefaultApiServerDatabaseConnectionPoolSize = 16
+
+  val DefaultMaxContractStateCacheSize: Long = 100000
```
Comment (miklos-da) [inline ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/ParticipantConfig.scala:39]: What dimension does this number have? Is this in bytes, megabytes or does it represent number of items or something else?

Comment (tudor-da) [reply ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/ParticipantConfig.scala:39]: It's maximum number of entries in the cache.

Comment (tudor-da) [reply ledger/participant-state/kvutils/app/src/main/scala/com/daml/ledger/participant/state/kvutils/app/ParticipantConfig.scala:39]: @miklos-da If you asked, it means it is ambiguous. I'll add a note to make it clear when we promote the feature from experimental to production-grade

**Pull request merged**